pkg/metrics/collector/netlink_metrics.go

```diff
@@ -15,85 +15,287 @@ package collector
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/GoogleCloudPlatform/netd/pkg/tcp_metrics/inetdiag"
+	"github.com/GoogleCloudPlatform/netd/pkg/tcp_metrics/parser"
+	"github.com/GoogleCloudPlatform/netd/pkg/tcp_metrics/tcp"
 )
 
 var (
 	TCPConnectionsDesc = prometheus.NewDesc(
-		"tcp_connections",
+		"active_tcp_connections",
 		"Current amount of tcp connections on a node",
-		nil, nil,
+		[]string{"pod_name", "namespace_name"}, nil,
 	)
 
 	ActiveTCPRetransmits = prometheus.NewDesc(
 		"active_tcp_retransmits",
 		"Current amount of tcp retransmits on a node",
-		nil, nil,
+		[]string{"pod_name", "namespace_name"}, nil,
 	)
 )
 
+type netlinkCollector struct {
+}
+
+type netlinkStats struct {
+	retransmits    uint64
+	tcpConnections uint64
+}
+
+type podDef struct {
+	name      string
+	namespace string
+}
+
+var (
+	mapMux   sync.RWMutex
+	ipMap    = make(map[string]podDef)
+	nodeName = os.Getenv("CURRENT_NODE_NAME")
+)
+
+func init() {
+
+	glog.Infof("Initializing netlink metrics")
+	err := createPodWatch()
+	if err != nil {
+		glog.Infof("Error, pod watch not running with err, %q", err)
+	}
+
+	registerCollector("netlink_metrics", NewNetlinkCollector)
+}
+
+// Functionality for multithreaded map usage
+func safeIPWrite(key string, value podDef) {
+	mapMux.Lock()
+	defer mapMux.Unlock()
+	ipMap[key] = value
+}
+
+func safeIPDelete(key string) {
+	mapMux.Lock()
+	defer mapMux.Unlock()
+	delete(ipMap, key)
+}
+
+func safeIPRead(key string) (podDef, bool) {
+	mapMux.RLock()
+	defer mapMux.RUnlock()
+	podDef, ok := ipMap[key]
+	return podDef, ok
+}
+
+// Functionality for pod watching
+func createPodWatch() error {
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return err
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	watchlist := cache.NewListWatchFromClient(
+		clientset.CoreV1().RESTClient(),
+		"pods",
+		v1.NamespaceAll,
+		fields.OneTermEqualSelector("spec.nodeName", nodeName),
+	)
+	_, controller := cache.NewInformer(
+		watchlist,
+		&v1.Pod{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    onPodAdd,
+			UpdateFunc: onPodUpdate,
+			DeleteFunc: onPodDelete,
+		},
+	)
+
+	stopper := make(chan struct{})
+
+	go controller.Run(stopper)
+
+	return nil
+}
+
+func podAllow(pod *v1.Pod) bool {
+	// Pod not assigned.
+	if pod.Status.PodIP == "" {
+		return false
+	}
+	// Pod running on host network.
+	if pod.Status.PodIP == pod.Status.HostIP {
+		return false
+	}
+	return true
+}
+
+func onPodAdd(podObj interface{}) {
+	pod := podObj.(*v1.Pod)
+	if podAllow(pod) {
+		safeIPWrite(pod.Status.PodIP, podDef{name: pod.ObjectMeta.Name, namespace: pod.ObjectMeta.Namespace})
+	}
+}
+
+func onPodUpdate(oldPod, newPod interface{}) {
+	prevPod := oldPod.(*v1.Pod)
+	pod := newPod.(*v1.Pod)
+	if prevPod.Status.PodIP == pod.Status.PodIP {
```

Comment:
I'm under the impression that pod IP is immutable so we don't need not to watch for pod updates?

Reply:
When a pod gets created the pod ip is sometimes not initialized yet. Once the pod is updated to ready, the ip gets added so there are a few cases where we need to watch for pod updates. I ran a test cluster with this configuration and it seemed to confirm this.

Reply:
Yes, we have to watch for pod updates. Kubelet only starts the CRI/CNI after the pod object exists.